File: Combinations - Leetcode 77/Combinations - Leetcode 77.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ans;
        vector<int> sol;
        backtrack(n, k, sol, ans);
        return ans;
    }

private:
    void backtrack(int x, int k, vector<int>& sol, vector<vector<int>>& ans) {
        if (sol.size() == k) {
            ans.push_back(sol);
            return;
        }

        if (x > k - sol.size()) {
            backtrack(x - 1, k, sol, ans);
        }

        sol.push_back(x);
        backtrack(x - 1, k, sol, ans);
        sol.pop_back();
    }
};

File: Combinations - Leetcode 77/Combinations - Leetcode 77.java
import java.util.*;

public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> sol = new ArrayList<>();
        backtrack(n, k, sol, ans);
        return ans;
    }

    private void backtrack(int x, int k, List<Integer> sol, List<List<Integer>> ans) {
        if (sol.size() == k) {
            ans.add(new ArrayList<>(sol));
            return;
        }

        if (x > k - sol.size()) {
            backtrack(x - 1, k, sol, ans);
        }

        sol.add(x);
        backtrack(x - 1, k, sol, ans);
        sol.remove(sol.size() - 1);
    }
}
